### PR TITLE
CSC RecHit Builder updates

### PR DESCRIPTION
**Pull request by nvoytish**
#### PR description:

This PR adds some features to the CSC RecHit Builder.

The addressed problems are:
- simple two overlapped signals delimitation in terms of the strip coordinate;
- improvement of the hit reconstruction in the extreme wire groups of chamber(best improvement for ME11);
- ME11 transition region hit reconstruction improvement.

For details see:
https://indico.cern.ch/event/968978/contributions/4078288/attachments/2133771/3594383/20_10_30-Voytishin_Palichik-CSC_rh_builder_improvement_updates.pdf

### CODE REVIEW

**Comment by nvoytish** [comment]
adding @ptcox as co-author

**Review by nvoytish — COMMENTED** [review]
code style improved as suggested in https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-32052/19639/code-format.patch

**Comment by nvoytish** [comment]
can somebody tell me where is seen the difference in the current version and proposed one in https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-32052/19642/code-format.patch  ?

**Comment by slava77** [comment]
> can somebody tell me where is seen the difference in the current version and proposed one in https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-32052/19642/code-format.patch ?

you can highlight the text to see the differences in whitespace

**Comment by perrotta** [comment]
please test

File: RecoLocalMuon/CSCRecHitD/src/CSCHitFromStripOnly.cc [hunk 1]
```diff
@@ -472,11 +469,69 @@ void CSCHitFromStripOnly::findMaxima(const CSCDetId& id) {
...
-}
-//
+
+      bool additional_maxima_found = false;
+      // search for additional maxima if:
+      // - hit is closer than 3 strips from the edge
+      // - enough consecutive strips with signal
+      // - strip charge distribution looks abnormal
+
+      if (i > 2 && i < (thePulseHeightMap.size() - 3) && numberOfConsecutiveStrips > 3) {
```
Comment (perrotta) [inline RecoLocalMuon/CSCRecHitD/src/CSCHitFromStripOnly.cc:479]: thePulseHeightMap.size() is an unsigned, as well as i . It is then safer avoiding differences with unsigned:
```suggestion
      if (i > 2 && i  +3< thePulseHeightMap.size() && numberOfConsecutiveStrips > 3) {
```

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoLocalMuon/CSCRecHitD/src/CSCHitFromStripOnly.cc [hunk 2]
```diff
@@ -472,11 +469,69 @@ void CSCHitFromStripOnly::findMaxima(const CSCDetId& id) {
...
+        //try to look for additional maxima at the left side from the main maxima
+
+        if (((thePulseHeightMap[i + 1].phmax() >= thePulseHeightMap[i - 1].phmax() &&
+              thePulseHeightMap[i + 1].phmax() >= thePulseHeightMap[i - 2].phmax() &&
+              thePulseHeightMap[i + 2].phmax() <= thePulseHeightMap[i - 2].phmax()) ||
+             (thePulseHeightMap[i + 1].phmax() <= thePulseHeightMap[i - 1].phmax() &&
+              thePulseHeightMap[i + 1].phmax() <= thePulseHeightMap[i - 2].phmax())) &&
+            thePulseHeightMap[i - 1].phmax() >=
+                thePulseHeightMap[i - 2]
+                    .phmax() &&  //to avoid close maxima delimitation (this is already present in the code)
```
Comment (slava77) [inline RecoLocalMuon/CSCRecHitD/src/CSCHitFromStripOnly.cc:489]:
```suggestion
            //to avoid close maxima delimitation (this is already present in the code)
            thePulseHeightMap[i - 1].phmax() >= thePulseHeightMap[i - 2].phmax() &&
```
to improve readability (unfortunately the clang-format breaks up the lines with code and comments; placing them on separate lines is to avoid the issue)

File: RecoLocalMuon/CSCRecHitD/src/CSCHitFromStripOnly.cc [hunk 3]
```diff
@@ -472,11 +469,69 @@ void CSCHitFromStripOnly::findMaxima(const CSCDetId& id) {
...
+            thePulseHeightMap[i - 2].phmax() > 20) {  //no need in a small charge maxima (might need adjustment)
+
+          additional_maxima_found = true;
+          theMaxima.push_back(i - 2);  //insert left maxima first
+          theConsecutiveStrips.push_back(
+              numberOfConsecutiveStrips);  //insert the same number of cosecutive strips, because they belong to both maximas
```
Comment (slava77) [inline RecoLocalMuon/CSCRecHitD/src/CSCHitFromStripOnly.cc:495]:
```suggestion
           //insert the same number of cosecutive strips, because they belong to both maximas
          theConsecutiveStrips.push_back(numberOfConsecutiveStrips);
```

File: RecoLocalMuon/CSCRecHitD/src/CSCHitFromStripOnly.cc [hunk 4]
```diff
@@ -472,11 +469,69 @@ void CSCHitFromStripOnly::findMaxima(const CSCDetId& id) {
...
+
+        //try to look for additional maxima at the right side from the main maxima
+        if (((thePulseHeightMap[i + 1].phmax() >= thePulseHeightMap[i - 1].phmax() &&
+              thePulseHeightMap[i + 2].phmax() >= thePulseHeightMap[i - 1].phmax()) ||
+             (thePulseHeightMap[i + 1].phmax() <= thePulseHeightMap[i - 1].phmax() &&
+              thePulseHeightMap[i + 2].phmax() <= thePulseHeightMap[i - 1].phmax() &&
+              thePulseHeightMap[i + 2].phmax() >= thePulseHeightMap[i - 2].phmax())) &&
+            thePulseHeightMap[i + 1].phmax() >=
+                thePulseHeightMap[i + 2]
+                    .phmax() &&  //to avoid close maxima delimitation (this is already present in the code)
```
Comment (slava77) [inline RecoLocalMuon/CSCRecHitD/src/CSCHitFromStripOnly.cc:510]: move the comments to a separate line.
Please update the lines below as well

File: RecoLocalMuon/CSCRecHitD/src/CSCHitFromStripOnly.cc [hunk 5]
```diff
@@ -472,11 +469,69 @@ void CSCHitFromStripOnly::findMaxima(const CSCDetId& id) {
...
+            thePulseHeightMap[i - 2].phmax() > 20) {  //no need in a small charge maxima (might need adjustment)
+
+          additional_maxima_found = true;
+          theMaxima.push_back(i - 2);  //insert left maxima first
+          theConsecutiveStrips.push_back(
+              numberOfConsecutiveStrips);  //insert the same number of cosecutive strips, because they belong to both maximas
+          theMaxima.push_back(i);          //insert main maxima
+          theConsecutiveStrips.push_back(
+              numberOfConsecutiveStrips);  //insert the same number of cosecutive strips, because they belong to both maximas
```
Comment (slava77) [inline RecoLocalMuon/CSCRecHitD/src/CSCHitFromStripOnly.cc:498]: move the comments to a separate line

File: RecoLocalMuon/CSCRecHitD/src/CSCMake2DRecHit.cc [hunk 6]
```diff
@@ -140,37 +140,44 @@ CSCRecHit2D CSCMake2DRecHit::hitFromStripAndWire(const CSCDetId& id,
...
+  wglo = layergeom_->middleWireOfGroup(wglo) + 0.5 - 0.5 * layergeom_->numberOfWiresPerGroup(wglo);
+  // Last wire in last wiregroup of cluster (OK if only 1 wiregroup too)
+  wghi = layergeom_->middleWireOfGroup(wghi) - 0.5 + 0.5 * layergeom_->numberOfWiresPerGroup(wghi);
 
   float stripWidth = -99.f;
   // If at the edge, then used 1 strip cluster only
   if (centerStrip == 1 || centerStrip == specs_->nStrips() || nStrip < 2) {
-    lp0 = layergeom_->stripWireIntersection(centerStrip, centerWire);
+    lp0 = (layergeom_->possibleRecHitPosition(float(centerStrip) - 0.5, wglo, wghi)).first;
+    ymiddle = lp0.y();
```
Comment (slava77) [inline RecoLocalMuon/CSCRecHitD/src/CSCMake2DRecHit.cc:156]: this leads to a dead assignment
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f461d2/10621/llvm-analysis/report-41ceff.html#EndPath
please check the code; if the `ymiddle` is not needed here, better remove it and perhaps even move its declaration from  L145 to the `else` below where it's actually used

File: RecoLocalMuon/CSCRecHitD/src/CSCMake2DRecHit.cc [hunk 7]
```diff
@@ -140,37 +140,44 @@ CSCRecHit2D CSCMake2DRecHit::hitFromStripAndWire(const CSCDetId& id,
...
-                                 centerStrip,
-                                 xWithinChamber,
-                                 stripWidth,
-                                 tpeak,
-                                 positionWithinTheStrip,
-                                 sigmaWithinTheStrip,
-                                 quality);
-    }
-    lp0 = LocalPoint(xWithinChamber, layergeom_->yOfWire(centerWire, xWithinChamber));
+    xMatchGatti_->findXOnStrip(id,
```
Comment (slava77) [inline RecoLocalMuon/CSCRecHitD/src/CSCMake2DRecHit.cc:170]: how frequently is this condition expected to fail?
The (removed) comment implies that the cost may be large.
If the cost is large, it would be better to find a way to not skip this if completely

File: RecoLocalMuon/CSCRecHitD/src/CSCRecHitDBuilder.cc [hunk 8]
```diff
@@ -124,8 +125,15 @@ void CSCRecHitDBuilder::build(const CSCStripDigiCollection* stripdc,
     for (auto const& s_hit : cscStripHit) {
       for (auto const& w_hit : cscWireHit) {
         CSCRecHit2D rechit = make2DHits_->hitFromStripAndWire(sDetId, layer, w_hit, s_hit);
-
-        bool isInFiducial = make2DHits_->isHitInFiducial(layer, rechit);
+        // Store rechit as a Local Point:
+        LocalPoint rhitlocal = rechit.localPosition();
+        float yreco = rhitlocal.y();
+        bool isInFiducial = false;
+        if ((sDetId.station() == 1) && (sDetId.ring() == 1 || sDetId.ring() == 4) && (fabs(yreco + 30.) < 5.)) {
```
Comment (slava77) [inline RecoLocalMuon/CSCRecHitD/src/CSCRecHitDBuilder.cc:132]: some comment about the origin of 30 and 5 would be useful.
Is it the +-5 cm from the A/B boundary?

**Comment by slava77** [comment]
enable profiling

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@nvoytish 
please clarify on the status of follow up to the code review in https://github.com/cms-sw/cmssw/pull/32052#pullrequestreview-529379286
Thank you.

File: RecoLocalMuon/CSCRecHitD/src/CSCMake2DRecHit.cc (hunk 7 above)
Comment (nvoytish) [reply RecoLocalMuon/CSCRecHitD/src/CSCMake2DRecHit.cc:170]: As far as I understand this we get rid of such cases when the wire+strip intersection gives us a local coordinate outside the layer's sensitive area (see slide 12 of presentation attached to this PR). That's why there is no need in this check. This is @ptcox suggestion for the code. If I am wrong, Tim, could you please comment?

File: RecoLocalMuon/CSCRecHitD/src/CSCRecHitDBuilder.cc (hunk 8 above)
Comment (nvoytish) [reply RecoLocalMuon/CSCRecHitD/src/CSCRecHitDBuilder.cc:133]: I added the required comment. You understood the idea correctly.

File: RecoLocalMuon/CSCRecHitD/src/CSCMake2DRecHit.cc (hunk 6 above)
Comment (nvoytish) [reply RecoLocalMuon/CSCRecHitD/src/CSCMake2DRecHit.cc]: `ymiddle` was used earlier in `if` and `else`, but not anymore. I forgat to move the declaration. fixed now.

File: RecoLocalMuon/CSCRecHitD/src/CSCHitFromStripOnly.cc (hunk 5 above)
Comment (nvoytish) [reply RecoLocalMuon/CSCRecHitD/src/CSCHitFromStripOnly.cc]: done

File: RecoLocalMuon/CSCRecHitD/src/CSCHitFromStripOnly.cc (hunk 4 above)
Comment (nvoytish) [reply RecoLocalMuon/CSCRecHitD/src/CSCHitFromStripOnly.cc]: done

**Comment by slava77** [comment]
@nvoytish 
please post a message when you are done with updates in this PR.

**Comment by nvoytish** [comment]
> @nvoytish
> please post a message when you are done with updates in this PR.

Hi, @slava77 ! If I didn't miss any issues, it looks like I am done with updates.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/32199	cf54157 
- updates are in line with the details presented in the reco meeting (see the link in the PR description)
- jenkins tests pass and comparisons with the baseline show differences starting from somewhat small changes in the cscSegments and propagating to downstream quantities. As expected, there are more segments and the segments are longer.

**Comment by qliphy** [comment]
+1

**Comment by makortel** [comment]
This PR might be causing heap-buffer-overflow failures in ASAN builds in `CSCXonStrip_MatchGatti::estimated2GattiCorrection()`, please see #32274.

File: RecoLocalMuon/CSCRecHitD/src/CSCMake2DRecHit.cc (hunk 7 above)
Comment (ptcox) [reply RecoLocalMuon/CSCRecHitD/src/CSCMake2DRecHit.cc:168]: Sorry I only just saw I'd been asked to comment.
Yes, this 'inside'  test was a sanity check to catch edge cases where wire and strip intersection fell outside the chamber.  It was only an issue for ME1/1 extreme wire groups. The new code has this sanity check implicitly included (in CSCLayerGeometry::intersectionOfStripAndWire) so the test is unnecessary.
